pkg/controller/resourcetree/tree.go

```diff
@@ -0,0 +1,130 @@
+package resourcetree
+
+import (
+	"context"
+	"github.com/oslokommune/okctl/pkg/api"
+)
+
+// ResourceNodeType defines what type of resource a ResourceNode represents
+type ResourceNodeType int
+const (
+	// ResourceNodeTypeGroup represents a node that has no actions associated with it. For now, only the root node
+	ResourceNodeTypeGroup ResourceNodeType = iota
+	// ResourceNodeTypeZone represents a HostedZone resource
+	ResourceNodeTypeZone
+	// ResourceNodeTypeVPC represents a VPC resource
+	ResourceNodeTypeVPC
+	// ResourceNodeTypeCluster represents a EKS cluster resource
+	ResourceNodeTypeCluster
+	// ResourceNodeTypeExternalSecrets represents an External Secrets resource
+	ResourceNodeTypeExternalSecrets
+	// ResourceNodeTypeALBIngress represents an ALB Ingress resource
+	ResourceNodeTypeALBIngress
+	// ResourceNodeTypeExternalDNS represents an External DNS resource
+	ResourceNodeTypeExternalDNS
+	// ResourceNodeTypeGithub represents a Github setup
+	ResourceNodeTypeGithub
+	// ResourceNodeTypeIdentityManager represents a Identity Manager resource
+	ResourceNodeTypeIdentityManager
+	// ResourceNodeTypeArgoCD represents an ArgoCD resource
+	ResourceNodeTypeArgoCD
+)
+
+// ResourceNodeState defines what state the resource is in, used to infer what action to take
+type ResourceNodeState int
+const (
+	// ResourceNodeStateNoop represents a state where no action is needed. E.g.: if the desired state of the
+	// resource conforms with the actual state
+	ResourceNodeStateNoop ResourceNodeState = iota
+	// ResourceNodeStatePresent represents the state where the resource exists
+	ResourceNodeStatePresent
+	// ResourceNodeStateAbsent represents the state where the resource does not exist
+	ResourceNodeStateAbsent
+)
+
+// CommonMetadata represents metadata required by most if not all operations on services
+type CommonMetadata struct {
+	Ctx context.Context
+	Id api.ID
+}
+
+// StateRefreshFn is a function that attempts to retrieve state potentially can only be retrieved at runtime. E.g.:
+// state that can only exist after an external resource has been created
+type StateRefreshFn func(node *ResourceNode)
+
+// ResourceNode represents a component of the cluster and its dependencies
+type ResourceNode struct {
+	Type  ResourceNodeType
+	State ResourceNodeState
+
+	// Contains metadata regarding the resource supplied by the desired state definition
```

Comment:
tree.go and this package should not anything abut "desired state", no? Consider rewording.